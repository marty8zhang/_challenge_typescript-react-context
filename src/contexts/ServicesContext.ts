import React from 'react';
import services from '../services';

const ServicesContext = React.createContext(services);

export { ServicesContext as default };
