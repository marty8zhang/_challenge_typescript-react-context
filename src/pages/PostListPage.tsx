import React from 'react';
import ServicesContext from '../contexts/ServicesContext';
import { PostGatewayInterface } from '../gateways/PostGateway';
import Post from '../models/Post';

export interface PostListPageProps {
  userId: number;
}

export interface PostListPageState {
  posts: Post[];
}

export default class PostListPage extends React.Component<PostListPageProps, PostListPageState> {
  constructor(props: Readonly<any>) {
    super(props);

    this.state = {
      posts: [],
    };
  }

  componentDidMount() {
    const { userId } = this.props;
    const { postGateway }: {postGateway: PostGatewayInterface} = this.context;

    postGateway.getPostsByUserId(userId)
      .then((posts: Post[]) => {
        this.setState({
          posts,
        });
      });
  }

  render() {
    const { posts } = this.state;
    const postList = posts.map((post) => (
      <tr key={post.id}>
        <td>{post.title}</td>
        <td>{post.body}</td>
      </tr>
    ));

    return (
      <table>
        <thead>
          <tr>
            <th>Title</th>
            <th>Page</th>
          </tr>
        </thead>
        <tbody>
          {postList.length ? postList
            : (
              <tr>
                <td colSpan={2}>No Post Found.</td>
              </tr>
            )}
        </tbody>
      </table>
    );
  }
}

PostListPage.contextType = ServicesContext;
