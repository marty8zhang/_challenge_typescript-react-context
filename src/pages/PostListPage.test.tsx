import React from 'react';
import { render, screen } from '@testing-library/react';
import { Route, Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import ServicesContext from '../contexts/ServicesContext';
import PostListPage from './PostListPage';

it('displays a table of the retrieved posts', async () => {
  const mockedGetPostsByUserId = jest.fn().mockResolvedValue([
    {
      userId: 1,
      id: 1,
      title: 'Test Post 1 Title',
      body: 'Test post 1 body',
    },
    {
      userId: 1,
      id: 2,
      title: 'Test Post 2 Title',
      body: 'Test post 2 body',
    },
  ]);
  const mockedServices = {
    postGateway: {
      getPostsByUserId: mockedGetPostsByUserId,
    },
  };

  const history = createMemoryHistory();
  history.push('/post/1');
  const { container } = render(
    <ServicesContext.Provider value={mockedServices}>
      <Router history={history}>
        <Route path="/post/:userId">
          <PostListPage />
        </Route>
      </Router>
    </ServicesContext.Provider>,
  );

  await screen.findByText('Test Post 1 Title');
  expect(mockedGetPostsByUserId).toBeCalledTimes(1);
  expect(mockedGetPostsByUserId).toBeCalledWith(1);

  const table = container.querySelector('table');
  expect(table).toMatchSnapshot();
});

it('displays a message in the table when no post is retrieved', async () => {
  const mockedGetPostsByUserId = jest.fn().mockResolvedValue([]);
  const mockedServices = {
    postGateway: {
      getPostsByUserId: mockedGetPostsByUserId,
    },
  };

  const history = createMemoryHistory();
  history.push('/post/2');
  const { container } = render(
    <ServicesContext.Provider value={mockedServices}>
      <Router history={history}>
        <Route path="/post/:userId">
          <PostListPage />
        </Route>
      </Router>
    </ServicesContext.Provider>,
  );

  const message = await screen.findByText('No Post Found.');
  expect(message).toBeInTheDocument();

  const table = container.querySelector('table');
  expect(table).toMatchSnapshot();
});
