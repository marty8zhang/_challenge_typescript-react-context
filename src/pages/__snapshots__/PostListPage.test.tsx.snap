// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays a message in the table when no post is retrieved 1`] = `
<table>
  <thead>
    <tr>
      <th>
        Title
      </th>
      <th>
        Page
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td
        colspan="2"
      >
        No Post Found.
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`displays a table of the retrieved posts 1`] = `
<table>
  <thead>
    <tr>
      <th>
        Title
      </th>
      <th>
        Page
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        Test Post 1 Title
      </td>
      <td>
        Test post 1 body
      </td>
    </tr>
    <tr>
      <td>
        Test Post 2 Title
      </td>
      <td>
        Test post 2 body
      </td>
    </tr>
  </tbody>
</table>
`;
